pkg/pubsubsource/pubsub_test.go

```diff
@@ -0,0 +1,258 @@
+	err := ensureTopicAndSubscription(setupCtx, pubsubClient, TopicID, subscriptionID)
+	assert.Nil(t, err)
+	cancelSetup()
+	messageCh := make(chan sourcer.Message, 20)
+	subscription := pubsubClient.Subscription(subscriptionID)
+	pubsubSource := NewPubSubSource(pubsubClient, subscription, MAX_EXTENSION_PERIOD)
+	ctx, cancel := context.WithCancel(context.Background())
+	publishChan := make(chan struct{})
+	defer cancel()
+	pubsubSource.StartReceiving(ctx)
+	go func() {
+		sendMessages(ctx, pubsubClient, TopicID, 100)
+		close(publishChan)
+	}()
+	pubsubSource.Read(ctx, mocks.ReadRequest{
+		CountValue: 5,
+		Timeout:    20 * time.Second,
+	}, messageCh)
+	assert.Equal(t, 5, len(messageCh))
+
+	// Try reading 4 more messages
+	// Since the previous batch didn't get acked, the data source shouldn't allow us to read more messages
+	// We should get 0 messages, meaning the channel only holds the previous 5 messages
+	pubsubSource.Read(ctx, mocks.ReadRequest{
+		CountValue: 4,
+		Timeout:    10 * time.Second,
+	}, messageCh)
+	assert.Equal(t, 5, len(messageCh)) // Expecting no new messages
+
+	// acknowledging the first batch
+	msg1 := <-messageCh
+	msg2 := <-messageCh
+	msg3 := <-messageCh
+	msg4 := <-messageCh
+	msg5 := <-messageCh
+	pubsubSource.Ack(ctx, mocks.TestAckRequest{
+		OffsetsValue: []sourcer.Offset{msg1.Offset(), msg2.Offset(), msg3.Offset(), msg4.Offset(), msg5.Offset()},
+	})
+
+	pubsubSource.Read(ctx, mocks.ReadRequest{
+		CountValue: 6,
+		Timeout:    10 * time.Second,
+	}, messageCh)
+	assert.Equal(t, 6, len(messageCh))
+	<-publishChan
+}
+
+func TestPubSubSource_PendingMessageEqualsBufferSize(t *testing.T) {
+	setupCtx, cancelSetup := context.WithCancel(context.Background())
+	err := ensureTopicAndSubscription(setupCtx, pubsubClient, TopicID, subscriptionID)
+	assert.Nil(t, err)
+	cancelSetup()
+	subscription := pubsubClient.Subscription(subscriptionID)
+	pubsubSource := NewPubSubSource(pubsubClient, subscription, MAX_EXTENSION_PERIOD)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	pubsubSource.StartReceiving(ctx)
+	sendMessages(ctx, pubsubClient, TopicID, 1400)
+	val := pubsubSource.Pending(context.Background())
+	assert.Equal(t, MAX_OUT_STANDING_MESSAGES, int(val)) // MAX_OUT_STANDING_MESSAGES is buffer size
+}
+
+func TestPubSubSource_Pending2(t *testing.T) {
+	setupCtx, cancelSetup := context.WithCancel(context.Background())
+	err := ensureTopicAndSubscription(setupCtx, pubsubClient, TopicID, subscriptionID)
+	assert.Nil(t, err)
+	cancelSetup()
+	subscription := pubsubClient.Subscription(subscriptionID)
+	pubsubSource := NewPubSubSource(pubsubClient, subscription, MAX_EXTENSION_PERIOD)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	pubsubSource.StartReceiving(ctx)
+	messageCh := make(chan sourcer.Message, 1000)
+	sendMessages(ctx, pubsubClient, TopicID, 2000)
+	val := pubsubSource.Pending(context.Background())
+	assert.Equal(t, 1000, int(val))
```

Comment:
Is this part duplicated with the test case above? Can we send 900 messages, which doesn't make the buffer full and verify after reading, the pending is 500?